File: 67.add-binary.cpp
/*
 * @lc app=leetcode id=67 lang=cpp
 *
 * [67] Add Binary
 */

#include <alloca.h>
#include <bits/stdc++.h>

#include <cassert>

using namespace std;

// @lc code=start
class Solution {
 public:
  string addBinary(string a, string b) {
    size_t a_len = a.length();
    size_t b_len = b.length();
    string ret;
    char carry = 0;
    for (uint32_t i = 1; i <= 10000; i++) {
      if (a_len < i && b_len < i && carry == 0) {
        break;
      }
      char a_val = (a_len >= i) ? (a[a_len - i] - '0') : 0;
      char b_val = (b_len >= i) ? (b[b_len - i] - '0') : 0;
      char sum = a_val + b_val + carry;
      carry = sum / 2;
      ret.insert(ret.begin(), (sum % 2) + '0');
    }
    cerr << "@@@" << ret << endl;
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  assert(s.addBinary("11", "1") == "100");
  assert(s.addBinary("1010", "1011") == "10101");
  assert(s.addBinary("0", "0") == "0");
  assert(s.addBinary("111", "111") == "1110");
  assert(s.addBinary(
             "111111111111111111111111111111111111111111111111111111111",
             "111111111111111111111111111111111111111111111111111111111") ==
         "1111111111111111111111111111111111111111111111111111111110");
  return 0;
}

File: 67.add-binary.rs
/*
 * @lc app=leetcode id=67 lang=rust
 *
 * [67] Add Binary
 */

struct Solution;

// @lc code=start
impl Solution {
    pub fn add_binary(a: String, b: String) -> String {
        let mut a_val: u128 = 0;
        let mut b_val: u128 = 0;
        for c in a.chars() {
            a_val = (a_val << 1) + (if c == '1' { 1 } else { 0 });
        }
        for c in b.chars() {
            b_val = (b_val << 1) + (if c == '1' { 1 } else { 0 });
        }
        format!("{:b}", a_val + b_val)
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::add_binary("11".to_string(), "1".to_string()),
        "100"
    );
}

#[test]
fn test2() {
    assert_eq!(
        Solution::add_binary("1010".to_string(), "1011".to_string()),
        "10101"
    );
}

#[test]
fn test3() {
    assert_eq!(
        Solution::add_binary("10100000100100110110010000010101111011011001101110111111111101000000101111001110001111100001101".to_string(), 
        "110101001011101110001111100110001010100001101011101010000011011011001011101111001100000011011110011".to_string()),
        "110111101100010011000101110110100000011101000101011001000011011000001100011110011010010011000000000");
}
